Feature: frontend task queue

shamoon opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

This PR proposes to add a frontend viewer for file consumption tasks that was inspired by [discussion](https://github.com/paperless-ngx/paperless-ngx/discussions/999#discussioncomment-2793701). Screenshots below.

This is a little basic at the moment and you can view this data in the Django admin but this has a couple advantages:

- Its not in the Django admin =)
- This surfaces completed / failed / started / queued attempts, they of course persist across browser sessions etc.
- Currently theres just the ability to 'dismiss' a task which just kinda marks it as "read".
- It's kind of a framework and could be expanded down the road, of course

A little technical info:

- Extends Django Q `Task` model with new `PaperlessTask` model
- Uses signals for when task is created i.e. added to queue (because it turns out a task isnt saved to relational DB until _after_ its done, which we dont want), executed and finally after execution
- New API endpoints `/tasks/` and `/acknowledge_tasks/`

<img width="1621" alt="Screen Shot 2022-06-06 at 1 36 05 AM" src="https://user-images.githubusercontent.com/4887959/172126869-1135e106-8aa7-4c49-8451-047a3844d2f4.png">
<img width="1621" alt="Screen Shot 2022-06-06 at 1 36 09 AM" src="https://user-images.githubusercontent.com/4887959/172126872-0df31253-96ca-4f7a-8b25-71346b9cd2a1.png">

Fixes #999

### To Do:

- [ ] Feedback!
- [x] Python tests
- [x] e2e tests

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] ~~I have made corresponding changes to the documentation as needed.~~
- [x] I have checked my modifications for any breaking changes.

shamoon committed:
PaperlessTask and consumption_tasks endpoint

shamoon committed:
Basic tasks frontend view

Update app-frame.component.ts

shamoon committed:
Basic tasks frontend view

Update app-frame.component.ts

shamoon committed:
Toggle functionality for tasks list

shamoon committed:
update post_save signal receiver

shamoon committed:
acknowledge_tasks endpoint & basic UI

Update tasks.service.ts

shamoon committed:
Better tasks api endpoint

shamoon committed:
Update frontend for reworked tasks api endpoint

qcasey commented:
This is an excellent start! I can't really comment on the code but I think this is a good discussion to have.

Off the top of my head, I have a couple suggestions to extend this in the future:

1. Give reason for failure in Failed Tasks (`Short result` I think)
2. Badge # of "unread" Failed Tasks next to the sidebar menu item
3. Confirmation dialog before Dismissing All (but I don't think a timeout/wait is necessary like multi-Document delete)
4. Ability to filter task func (or remove non-document related tasks entirely). My Completed list is mostly `paperless_mail.tasks.process_mail_accounts`

I'm sure you have some of your own additions in mind. However this PR as-is creates much more visibility for the user and I'm excited for it.

shamoon commented:
> 4. Ability to filter task func (or remove non-document related tasks entirely). My Completed list is mostly `paperless_mail.tasks.process_mail_accounts`

those are cool ideas, thanks! I’ll work on them. One thing is regarding point 4, it should  have included _only_ file consumption tasks in the migration, sounds like there may be an issue there for me to fix.

shamoon committed:
Task results popover

fix merge

shamoon committed:
Task results popover

shamoon committed:
fix merge

shamoon committed:
Merge branch 'feature-frontend-task-queue' of github.com:paperless-ngx/paperless-ngx into feature-frontend-task-queue

shamoon committed:
Revert "Merge branch 'feature-frontend-task-queue' of github.com:paperless-ngx/paperless-ngx into feature-frontend-task-queue"

This reverts commit 011164bc32e639189bf9c7be0665199b61734e77, reversing
changes made to 0a43ce9cedf5158f21bb6f9ecc6ba2fdab205a1e.

shamoon committed:
Update 1021_paperlesstask.py

shamoon committed:
mobile improvements, fix dismiss 1 task

shamoon committed:
tasks usability improvements

shamoon committed:
show failed tasks badge in sidebar

shamoon committed:
fix result on empty task

shamoon committed:
automatically refresh tasks on file socket messages too

shamoon committed:
Simple e2e test for Tasks component

shamoon commented:
Cool, useful changes I think thanks, I added them (and some other things). Screenshots updated. Welcome more feedback!

I also added frontend tests, TBH I dont think I have the Django skills to do the backend tests for this. Its mostly I dont know how to mock the Q tasks and stuff like that. If anyone is willing help getting that part (or more) I can write more tests from there.

<img width="1785" alt="Screen Shot 2022-05-27 at 12 44 45 AM" src="https://user-images.githubusercontent.com/4887959/170656484-cdf9181b-2319-4ee9-8137-9325f14d4fb1.png">
<img width="1785" alt="Screen Shot 2022-05-27 at 12 55 27 AM" src="https://user-images.githubusercontent.com/4887959/170656493-3da138b0-1bdd-497f-a4b7-4a2913b47e13.png">

tooomm commented:
>However this PR as-is creates much more visibility for the user and I'm excited for it.

I do fully agree! Looking forward to this.

My thoughts:
- `Processing` is a better fit than `Incomplete`?
- You changed the tab order. Any particular reason? I think I liked the first more. :)
- Is it possible to show waiting/queued up documents in their own tab as well?
  When importing many at once, it gives an estimation on when things are finished and allows to better follow and understand the import and consumption process.

shamoon committed:
add collapsible task result output

shamoon committed:
Update e2e tests for collapsible result

shamoon commented:
Any @paperless-ngx/backend (or anyone else!) can help with the python tests here?

tooomm commented:
>I know what you mean but we can’t differentiate between Tasks that are actually in progress vs just haven’t started yet (ie queued) so in some ways incomplete is more accurate, I think?

Oh, in that case it's probably fine. I can not come up with a better term right now.

>I did! I guess I was thinking folks would be going here to view failed tasks more often than anything else. Especially if we keep the badge with the number of failed tasks in the sidebar. Open to changing this back of course if others prefer how it was before too.

👍 Just keep it.

>Apologies if I’m misunderstanding but the “Incomplete” tab should do this already no? Ie it shows all docs that are uploaded (or seen in consume folder) and in the queue for processing.

I thought, `Incomplete` has only documents in progress and not also the waiting ones, that's why I suggested the rename above.
As you mentioned, if we can not differentiate between current files in progress and ones waiting to be processed, there is no point in my request.

I'm just going by your screenshots. Thanks for your patient explanations. 😃 

<br>

I think there is a little graphic issue when the logs are unfolded? Right above the next `Dismiss` button on your screenshot.

shamoon commented:
> I think there is a little graphic issue when the logs are unfolded? Right above the next `Dismiss` button on your screenshot.

Thanks good catch =)

shamoon committed:
Use pre_execute signal to specify started vs queued

shamoon commented on src/documents/signals/handlers.py:
```diff
@@ -499,3 +501,25 @@ def add_to_index(sender, document, **kwargs):
+        paperless_task.save()
+
+
+@receiver(models.signals.post_save, sender=django_q.tasks.Task)
```
Ah cool yea that makes more sense but weirdly I cant get this loaded. Official docs show v1.3.6 which doesnt have this signal ( https://django-q.readthedocs.io/en/latest/signals.html ) but then this repo https://github.com/Koed00/django-q/blob/master/docs/signals.rst shows it but when I tried installing 1.3.9 I still dont see it (i.e. just get no attribute `post_execute`). Huh?

<img width="264" alt="Screen Shot 2022-06-06 at 12 15 27 AM" src="https://user-images.githubusercontent.com/4887959/172113656-365ca814-f419-4651-b7b7-9378b8b88b5b.png">

stumpylog replied:
Ah, I see the issue.  It's only in the `master` and never got released.  So I think the `post_save` is the best possible now, until #1014 goes in.

shamoon replied:
So in actually looking at this after rebasing on dev I think `post_save` is still more helpful to use because that gives us the actual `Task` model in the database as a `OneToOneField` which allows us quicker access to all the data that object already holds, like `paperless_task.attempted_task.success` etc

What do you think?

shamoon commented on src/documents/serialisers.py:
```diff
@@ -595,3 +596,62 @@ def create(self, validated_data):
+    status = serializers.SerializerMethodField()
+
+    def get_status(self, obj):
+        if obj.attempted_task is None:
```
Cool yea! I went with "Queued", "Started", "Completed", "Failed"

shamoon committed:
Merge branch 'dev' into feature-frontend-task-queue

shamoon committed:
Update tasks e2e tests to cypress 10

and remove debug print

shamoon committed:
allow PaperlessTask.started null

shamoon committed:
python tests for tasks api endpoints

tooomm commented:
The improved status is great! 👍 

Probably a little typo: "Complete**d**" to align with the other tabs

shamoon committed:
Merge branch 'dev' into feature-frontend-task-queue

shamoon committed:
fix conflicting migrations

bauerj commented on src/documents/serialisers.py:
```diff
@@ -592,3 +593,65 @@ def create(self, validated_data):
+        count = PaperlessTask.objects.filter(id__in=tasks).count()
+        if not count == len(tasks):
+            raise serializers.ValidationError(
+                f"Some tasks in {name} don't exist or were " f"specified twice.",
```
Is this string broken up by accident?

shamoon replied:
Think that was just a merge thing, thanks!

bauerj reviewed (approved):
Backend code looks great! Will also test this.
